### PR TITLE
#1521: Configure which fields are indexed in the Ditto search index per namespace pattern

### PR DESCRIPTION
**Pull request by an1310**
Long time coming.  

Resolves: #1521

### CODE REVIEW

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Thanks a lot @an1310 for that awesome PR.
Looking already really good - I did a quick round of comments to what I found.

What is still missing before getting merged:
* an "empty" default configuration in the `search.conf` file
   * containing e.g. commented out example of how the config for a concrete namespace would look like
* documentation on how to enable the `SearchIndexingSignalEnrichmentFacadeProvider`, maybe both:
   * in `search.conf` as commented out alternative to the default `DittoCachingSignalEnrichmentFacadeProvider` for `ditto.extensions.caching-signal-enrichment-facade-provider`
   * and as a subsection in the [Operating Ditto](https://eclipse.dev/ditto/installation-operating.html) docs
      * Next to [Restricting entity creation](https://eclipse.dev/ditto/installation-operating.html#restricting-entity-creation)
      * Source is here: [installation-operating.md](https://github.com/eclipse-ditto/ditto/edit/master/documentation/src/main/resources/pages/ditto/installation-operating.md)
      * Similar description as for the "Restricting entity creation"


And one _optional_ wish from me 😁 
I just figured out that the `entity-creation` configuration already supports adding wildcards.
So maybe we could integrate that in the same way, see:
https://github.com/eclipse-ditto/ditto/blob/2a35c6f60d22037f8fa3124806863ec1e5c1a8e6/policies/enforcement/src/main/java/org/eclipse/ditto/policies/enforcement/config/DefaultCreationRestrictionConfig.java#L37-L55

So basically the `getNamespace` would return a `Pattern` instead which is created during the config parsing.

An additional benefit of that would be that is configured commonly as the "Restriction entity creation".

What do you think about that?

File: thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfig.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+    private DefaultNamespaceSearchIndexConfig(final String namespace, final Collection<String> fields) {
+        this.namespace = namespace;
+        this.searchIncludeFields = Collections.unmodifiableList(new ArrayList<>(fields));
+    }
+
+    /**
+     * Returns an instance of {@code DefaultNamespaceSearchIndexConfig} based on the settings of the specified Config.
+     *
+     * @param namespace the namespace config passed in the {@code config}.
```
Comment (thjaeckle) [inline thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfig.java:60]: this `@param namespace` must be removed - there is no parameter named in that way.
That's also why the build currently is failing (due to Javadoc error).

File: thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfig.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    public static DefaultNamespaceSearchIndexConfig of(
+            final String namespace,
+            final Collection<String> fields) {
+        checkNotNull(namespace, "namespace");
+        checkNotEmpty(namespace, "namespace");
+        argumentNotEmpty(fields, "fields");
+
+        return new DefaultNamespaceSearchIndexConfig(namespace, fields);
+    }
```
Comment (thjaeckle) [inline thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfig.java:86]: This method is never called.
Propably a leftover from copy&paste - please remove it (also the Javadoc is wrong).

**Comment by thjaeckle** [comment]
Feature wise it works really great 👍 
I just manually tested and only the defined fields were inserted into the search index, awesome 🏅

File: internal/models/signalenrichment/src/main/java/org/eclipse/ditto/internal/models/signalenrichment/SignalEnrichmentCacheKey.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
  * Implementation for a {@link CacheKey} used in scope of signal enrichment.
  */
 @Immutable
-final class SignalEnrichmentCacheKey implements CacheKey<SignalEnrichmentContext> {
+public final class SignalEnrichmentCacheKey implements CacheKey<SignalEnrichmentContext> {
```
[the remark this answers is not in the document]
Comment (an1310) [reply internal/models/signalenrichment/src/main/java/org/eclipse/ditto/internal/models/signalenrichment/SignalEnrichmentCacheKey.java:30]: Ah, I missed this in my cleanup.

File: thingsearch/service/src/test/java/org/eclipse/ditto/thingsearch/service/starter/config/DittoSearchConfigTest.java [hunk 4]
```diff
@@ -61,6 +62,9 @@ public void testHashCodeAndEquals() {
     public void testQueryPersistenceConfig() {
         final var config = ConfigFactory.load("search-test.conf");
         final var underTest = DittoSearchConfig.of(DefaultScopedConfig.dittoScoped(config));
+
+        //assertThat(underTest.getSearchIncludeFields().isEmpty()).isTrue();
```
[the remark this answers is not in the document]
Comment (an1310) [reply thingsearch/service/src/test/java/org/eclipse/ditto/thingsearch/service/starter/config/DittoSearchConfigTest.java:66]: Old working copy -- I ended up adding a new unit test.

**Comment by an1310** [comment]
> Thanks a lot @an1310 for that awesome PR. Looking already really good - I did a quick round of comments to what I found.
> 
> What is still missing before getting merged:
> 
> * an "empty" default configuration in the `search.conf` file
>   
>   * containing e.g. commented out example of how the config for a concrete namespace would look like
> * documentation on how to enable the `SearchIndexingSignalEnrichmentFacadeProvider`, maybe both:
>   
>   * in `search.conf` as commented out alternative to the default `DittoCachingSignalEnrichmentFacadeProvider` for `ditto.extensions.caching-signal-enrichment-facade-provider`
>   * and as a subsection in the [Operating Ditto](https://eclipse.dev/ditto/installation-operating.html) docs
>     
>     * Next to [Restricting entity creation](https://eclipse.dev/ditto/installation-operating.html#restricting-entity-creation)
>     * Source is here: [installation-operating.md](https://github.com/eclipse-ditto/ditto/edit/master/documentation/src/main/resources/pages/ditto/installation-operating.md)
>     * Similar description as for the "Restricting entity creation"
> 
> And one _optional_ wish from me 😁 I just figured out that the `entity-creation` configuration already supports adding wildcards. So maybe we could integrate that in the same way, see:
> 
> https://github.com/eclipse-ditto/ditto/blob/2a35c6f60d22037f8fa3124806863ec1e5c1a8e6/policies/enforcement/src/main/java/org/eclipse/ditto/policies/enforcement/config/DefaultCreationRestrictionConfig.java#L37-L55
> 
> So basically the `getNamespace` would return a `Pattern` instead which is created during the config parsing.
> 
> An additional benefit of that would be that is configured commonly as the "Restriction entity creation".
> 
> What do you think about that?

I think this suggestion is great.  If it's okay, can we do this in a separate issue and I can contribute it?

File: thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/persistence/write/streaming/SearchIndexingSignalEnrichmentFacadeProvider.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                // Extract the first element
+                CharSequence pointerString = searchIncludeFields.get(0);
+
+                // Prepare the rest of the elements for varargs
+                CharSequence[] furtherPointerStrings = new CharSequence[searchIncludeFields.size() - 1];
+                for (int i = 1; i < searchIncludeFields.size(); i++) {
+                    furtherPointerStrings[i - 1] = searchIncludeFields.get(i);
+                }
+
+                JsonFieldSelector indexedFields = JsonFieldSelector.newInstance(pointerString, furtherPointerStrings);
```
[the remark this answers is not in the document]
Comment (an1310) [reply thingsearch/service/src/main/java/org/eclipse/ditto/thingsearch/service/persistence/write/streaming/SearchIndexingSignalEnrichmentFacadeProvider.java:94]: There's a function to create a JsonFieldSelector with an iterable of Strings with JsonParseOptions.  Is that the one you mean?

File: thingsearch/service/src/test/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfigTest.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        // Second config
+        softly.assertThat(second.getNamespace()).isEqualTo("org.eclipse.test");
+
+        softly.assertThat(second.getSearchIncludeFields())
+                .as(NamespaceSearchIndexConfig.NamespaceSearchIndexConfigValue.SEARCH_INCLUDE_FIELDS.getConfigPath())
+                .isEqualTo(
+                        List.of(
+                                "attributes",
+                                "features/info/properties/",
+                                "features/info/other"                        )
```
[the remark this answers is not in the document]
Comment (an1310) [reply thingsearch/service/src/test/java/org/eclipse/ditto/thingsearch/service/common/config/DefaultNamespaceSearchIndexConfigTest.java:100]: When I was researching formatting rules, I followed the example set in the org.eclipse.ditto.gateway.service.util.config.security.DefaultOauthConfigTest.  But I'll make this a bit more inline with everything else.

**Comment by thjaeckle** [comment]
One TODO to check (I can also check - but maybe you are quicker):
* what happens when you search for a field which is not indexed?
   * do you find nothing (I assume)
   * is there an error (there probably could be - and maybe even should be) indicating that this field is not indexed?

**Review by thjaeckle — APPROVED** [review]
@an1310 I did a review commit, hope you don't mind - providing:
* added a cache to SearchIndexingSignalEnrichmentFacade in order to only evaluate "patterns" once for a given namespace
* removed copy&pasted unit tests in SearchIndexingSignalEnrichmentFacadeTest by adding another abstract test class AbstractCachingSignalEnrichmentFacadeTest
* added a unit test testing selection of JsonFieldSelectors based on different namespaces
* minor cleanup and formatting
* enhanced documentation about how to configure the indexed namespaces via system properties

For me, this would now good to be merged 👍

**Comment by an1310** [comment]
I don't mind in the slightest. Glad to have helped.

**Comment by thjaeckle** [comment]
@an1310 I also just provided configuration via the Helm chart - so that it is easily enabled and configured via `values.yaml`

Once the build is green (nowadays GitLab runners seem quite busy), I will merge :)

Thanks a lot for this contribution - I guess not only you will benefit from that configuration.
